File: skeleton/src/me/shkschneider/skeleton/network/WebService.java
package me.shkschneider.skeleton.network;

import com.koushikdutta.async.future.Future;
import com.koushikdutta.async.future.FutureCallback;
import com.koushikdutta.ion.Response;
import me.shkschneider.skeleton.helpers.LogHelper;

public abstract class WebService<T> {

    public int id;
    public String url;
    public int tries;

    protected Future<Response<T>> request;

    private Callback callback;

    public WebService(final int id, final String url, final Callback callback) {
        this.id = id;
        this.url = url;
        this.tries = 0;
        this.request = null;
        this.callback = callback;
    }

    public WebService(final int id, final String url) {
        this(id, url, null);
    }

    public void run(final Callback overloadedCallback) {
        this.request.setCallback(new FutureCallback<Response<T>>() {

            @Override
            public void onCompleted(final Exception e, final Response<T> result) {
                if (e != null) {
                    LogHelper.error(e.getMessage());
                    if (overloadedCallback != null) {
                        overloadedCallback.webserviceFailure(e);
                    }
                    return;
                }

                if (overloadedCallback != null) {
                    overloadedCallback.webserviceSuccess(result.getHeaders().getResponseCode(), result.getResult());
                }
            }

        });
    }

    public void run() {
        run(this.callback);
    }

    public void failure(final Exception e) {
        if (this.callback != null) {
            this.callback.webserviceFailure(e);
        }
    }

    public void success(final int responseCode, final Object result) {
        if (this.callback != null) {
            this.callback.webserviceSuccess(responseCode, result);
        }
    }

    public boolean cancel() {
        return this.request.cancel();
    }

    public static interface Callback {

        public void webserviceFailure(final Exception e);

        public void webserviceSuccess(final int responseCode, final Object result);

    }

}

File: android/src/main/kotlin/me/shkschneider/skeleton/android/network/WebService.kt
package me.shkschneider.skeleton.android.network

import com.github.kittinunf.fuel.Fuel
import com.github.kittinunf.fuel.core.FuelError
import com.github.kittinunf.fuel.core.Request
import com.github.kittinunf.fuel.core.Response
import com.github.kittinunf.result.Result
import com.google.gson.Gson
import me.shkschneider.skeleton.kotlin.log.Logger

typealias WebServiceSuccess<T> = (Request, Response, T?) -> Unit
typealias WebServiceFailure = (Request, Response, Exception) -> Unit

@Suppress("DEPRECATION")
open class WebService(gson: Gson = Gson()) : BaseWebService(gson) {

    inline fun <reified T : Any> get(url: String, crossinline success: WebServiceSuccess<T>, noinline failure: WebServiceFailure? = null): Request =
        request<T>(Fuel.get(url), success, failure)

    inline fun <reified T : Any> head(url: String, crossinline success: WebServiceSuccess<T>, noinline failure: WebServiceFailure? = null): Request =
        request<T>(Fuel.head(url), success, failure)

    inline fun <reified T : Any> post(url: String, crossinline success: WebServiceSuccess<T>, noinline failure: WebServiceFailure? = null): Request =
        request<T>(Fuel.post(url), success, failure)

    inline fun <reified T : Any> put(url: String, crossinline success: WebServiceSuccess<T>, noinline failure: WebServiceFailure? = null): Request =
        request<T>(Fuel.put(url), success, failure)

    inline fun <reified T : Any> delete(url: String, crossinline success: WebServiceSuccess<T>, noinline failure: WebServiceFailure? = null): Request =
        request<T>(Fuel.delete(url), success, failure)

    @Deprecated("@hide")
    inline fun <reified T : Any> request(request: Request, crossinline success: WebServiceSuccess<T>, noinline failure: WebServiceFailure? = null): Request =
        request.responseObject(Deserializer(gson, T::class)) { _, response, result: Result<T, FuelError> ->
            result.fold({ data ->
                success(request, response, data)
            }, { fuelError ->
                Logger.error("${fuelError.response.statusCode} ${fuelError.response.url}", fuelError.exception)
                failure?.invoke(request, response, fuelError.exception)
            })
        }

}
